Add deprecation warnings for features to be removed/changed in RSpec 4

pirj opened the pull request:
A few things are a bit off, e.g.:
 - no deprecation for multi-condition filtering usages 🤷 
 - no specs for `:if`/`:unless` deprecation usages
 - had to *flip* the `unless` condition in `snippet_extractor_spec.rb` to skip it on Ruby 1.8 builds. Puzzled to understand how `:unless => RSpec::Support::Ruby.jruby? && !RUBY_VERSION.start_with?('1.8.')` was working ([example failure](https://github.com/rspec/rspec-core/runs/2110374357))
 - should we add deprecation warnings when running on rbx/ree/Ruby < 2.3?

This is purposed to be released as version 3.99.
Sibling PRs:
 - https://github.com/rspec/rspec-mocks/pull/1418
 - https://github.com/rspec/rspec-expectations/pull/1301

Release strategy: https://github.com/rspec/rspec-core/pull/2880#issuecomment-797756681
RSpec 4 plan: https://github.com/rspec/rspec/issues/61
Changes: https://github.com/rspec/rspec-core/blob/4-0-dev/Changelog.md#development

## TODO

 - [ ] merge this PR, then `rspec-expectations` PR, then `rspec-mocks` PR
 - [ ] revert [this](https://github.com/rspec/rspec-core/pull/2880/files#diff-89eebfcbc0f14b6d989517837ca1e94fce4e2ce9a03233641cd936f2b8d2ed94R84) `spec/spec_helper.rb` change

pirj committed:
Use :skip instead of :if/:else in specs

Related to https://github.com/rspec/rspec-core/issues/1821

pirj committed:
Add RSpec 4 deprecation warnings

See https://github.com/rspec/rspec/issues/61

pirj commented:
@JonRowe Am I doing the right thing?

JonRowe commented:
@pirj yep, thats the exactly the kind of thing we'll want in a final 3.x release.

We'll follow the 2.x strategy of:
- Release a 3.10 patch of any final bug fixes
- Release a final minor version containing any final enchancements
- Release 3.99 containing these deprecations
- Release 4.x

It's sufficient to keep these as a PR until we get to releasing 3.99, the first two steps are just release management as usual

pirj committed:
Adjust specs to be deprecation warning-free

pirj committed:
Clamp failing JRuby/Ruby 1.8 example

pirj commented:
Green!

I have seen `features/command_line/init.feature:25`, and `features/command_line/bisect.feature` (on JRuby) failing intermittently on CI. Might be flaky. Couldn't reproduce locally.

pirj committed:
Temporarily (for 3.99) prevent deprecation failures

See
https://github.com/rspec/rspec-mocks/pull/1418/checks?check_run_id=2115367793
for an example for RSpec Mocks build failure

benoittgt commented:
Thanks @pirj for this PR. 🙌🏻

> should we add deprecation warnings when running on rbx/ree/Ruby < 2.3?

It means people will always have the deprecations when they use the last version chosen by bundler for their Ruby version?

benoittgt commented:
@pirj are we good to merge it?

pirj commented on spec/spec_helper.rb:
```diff
@@ -80,8 +80,10 @@ def handle_current_dir_change
-  c.raise_errors_for_deprecations!
-  c.color = true
+
+  # Temporary measure to prevent failures when run in a sub-build from other repos
+  # c.raise_errors_for_deprecations!
```
This needs to be reverted when all three are merged.

JonRowe commented:
Migrated to the monorepo
